### DISCUSSION TITLE
How to return a join query from a function?

### DISCUSSION THREAD

**Discussion by cometw** [diesel-rs/diesel#3743] [Q&A] [answered]
I have a query that is repeated a lot throughout the project, that essentially narrows down the domain for possible search.

The query is fairly large so I am trying to break it down into type-able parts.

Right now I'm trying to get this working:

``` rust
use diesel::{
    alias,
    helper_types::{AliasedFields, Eq, LeftJoinOn, LeftJoinQuerySource},
    query_source::Alias,
    BoolExpressionMethods, ExpressionMethods, JoinOnDsl, QueryDsl, SelectableHelper,
};

alias!(
    visibility as team_visibility_alias: TeamVisibilityAlias,
    visibility as entity_visibility_alias: EntityVisibilityAlias
);

pub(crate) fn narrow() -> LeftJoinQuerySource<
    LeftJoinQuerySource<
        admin::table,
        Alias<TeamVisibilityAlias>,
        Eq<AliasedFields<TeamVisibilityAlias, visibility::object>, admin::team_id>,
    >,
    Alias<EntityVisibilityAlias>,
    Eq<AliasedFields<EntityVisibilityAlias, visibility::object>, admin::entity_id>,
> {
    admin::table
        .left_join(
            team_visibility_alias.on(team_visibility_alias
                .field(visibility::object)
                .eq(admin::team_id)),
        )
        .left_join(
            entity_visibility_alias.on(entity_visibility_alias
                .field(visibility::object)
                .eq(admin::entity_id)),
        )
}
```
As far as I can tell, the type seems correct, but the compiler tells me otherwise:

``` sh
E0308  mismatched types
 expected struct `JoinOn<Join<JoinOn<Join<table, Alias<TeamVisibilityAlias>, LeftOuter>, Grouped<Eq<..., ...>>>, ..., ...>, ...>`
    found struct `SelectStatement<FromClause<JoinOn<Join<JoinOn<Join<table, Alias<...>, ...>, ...>, ..., ...>, ...>>>`
```
Essentially the type is _**almost**_ there, except for this part
```
SelectStatement<FromClause<[expected-type]>>
```
Although I don't quite see where that `SelectStatement` came from. I think I'm missing something here, perhaps `admin::table` is itself of some type that on joins builds up to `SelectStatement<FromClause<...>>`. If so what `helper_type` should I substitute here?

**Comment by weiznich** [comment, marked as answer]
The problem here is that you return a `LeftJoinQuerySource` instead of just `LeftJoin<…>` as shown in the documentation of [`left_join`](https://docs.diesel.rs/2.1.x/diesel/prelude/trait.QueryDsl.html#method.left_join) (note the return type). Changing both occurrences of `LeftJoinQuerySource` to `LeftJoin` should fix the problem. 

As an additional note: You might be interested in https://github.com/diesel-rs/diesel/pull/3683, which should simplify these kind of pattern drastically. It's only available on the master branch for now, but will be included in the next feature release.

**Reply by cometw** [reply]
I knew I was missing something obvious! That's fixed it, thanks. 

Also #3683 is exactly what I was looking for! It'd save me a ton of time
